symbolic-base/src/ZkFold/Data/Vector.hs

```diff
@@ -160,9 +164,34 @@ concat = Vector . V.concatMap toV . toV
 unsafeConcat :: forall m n a . [Vector n a] -> Vector (m * n) a
 unsafeConcat = concat . unsafeToVector @m
 
+-- | Map a function over a vector and concatenate the results.
+concatMap :: forall m n a b . (a -> Vector n b) -> Vector m a -> Vector (m * n) b
+concatMap f (Vector v) = Vector $ V.concatMap (toV . f) v
+
+-- | Pair each element of a vector with its index.
+indexed :: forall n a. KnownNat n => Vector n a -> Vector n (Natural, a)
+indexed = mapWithIx (,)
+
 chunks :: forall m n a . KnownNat n => Vector (m * n) a -> Vector m (Vector n a)
 chunks (Vector vectors) = unsafeToVector (Vector <$> V.chunksOf (fromIntegral $ value @n) vectors)
 
+-- | Slice a vector of size @size@, starting at index @i@, and taking @n@ elements.
+--
+-- Note that we'll get run-time error if @i + n > size@.
+slice :: forall i n size a.
+  KnownNat i
+  => KnownNat n
+  => Vector size a
+  -> Vector n a
+slice (Vector v) = Vector $ V.slice (fromIntegral $ value @i) (fromIntegral $ value @n) v
+
+-- | Yield the vector obtained by replacing each element @i@ of the
+-- index vector by @xs'!'i@.
+--
+-- Note that we'll get run-time error if any index in the index vector is out of bounds.
```

Comment:
If you used `Zp n` instead of `Natural` here, there wouldn't be out-of-bounds indices